jetty-core/jetty-ee/src/main/java/org/eclipse/jetty/ee/WebappProtectedClasses.java

```diff
@@ -0,0 +1,129 @@
+     * @param patterns the patterns to use
+     */
+    public static void addSystemClasses(String... patterns)
+    {
+        DEFAULT_SYSTEM_CLASSES.add(patterns);
+    }
+
+    /**
+     * Add a System Class pattern to use for all WebAppContexts of a given {@link Server}.
+     * @param server The {@link Server} instance to add classes to
+     * @param patterns the patterns to use
+     */
+    public static void addSystemClasses(Server server, String... patterns)
+    {
+        addClasses(server, SYSTEM_CLASSES_ATTRIBUTE, DEFAULT_SYSTEM_CLASSES, patterns);
+    }
+
+    /**
+     * Add a System Class pattern to use for WebAppContexts of a given environment.
+     * @param environment The {@link Environment} instance to add classes to
+     * @param patterns the patterns to use
+     */
+    public static void addSystemClasses(Environment environment, String... patterns)
+    {
+        addClasses(environment, SYSTEM_CLASSES_ATTRIBUTE, DEFAULT_SYSTEM_CLASSES, patterns);
+    }
+
+    public static ClassMatcher getServerClasses(Server server)
+    {
+        return getClassMatcher(server, SERVER_CLASSES_ATTRIBUTE, DEFAULT_SERVER_CLASSES);
+    }
+
+    public static ClassMatcher getServerClasses(Environment environment)
+    {
+        return getClassMatcher(environment, SERVER_CLASSES_ATTRIBUTE, DEFAULT_SERVER_CLASSES);
+    }
+
+    /**
+     * Add a Server Class pattern to use for all WebAppContexts of a given {@link Server}.
+     * @param patterns the patterns to use
+     */
+    public static void addServerClasses(String... patterns)
+    {
+        DEFAULT_SERVER_CLASSES.add(patterns);
+    }
+
+    /**
+     * Add a Server Class pattern to use for all WebAppContexts of a given {@link Server}.
+     * @param server The {@link Server} instance to add classes to
+     * @param patterns the patterns to use
+     */
+    public static void addServerClasses(Server server, String... patterns)
+    {
+        addClasses(server, SERVER_CLASSES_ATTRIBUTE, DEFAULT_SERVER_CLASSES, patterns);
+    }
+
+    /**
+     * Add a Server Class pattern to use for all ee9 WebAppContexts.
+     * @param environment The {@link Environment} instance to add classes to
+     * @param patterns the patterns to use
+     */
+    public static void addServerClasses(Environment environment, String... patterns)
+    {
+        addClasses(environment, SERVER_CLASSES_ATTRIBUTE, DEFAULT_SERVER_CLASSES, patterns);
+    }
+
+    private static void addClasses(Attributes attributes, String attribute, ClassMatcher defaultPatterns, String... patterns)
+    {
+        ClassMatcher classMatcher = getClassMatcher(attributes, attribute, defaultPatterns);
+        if (patterns != null && patterns.length > 0)
+            classMatcher.add(patterns);
+    }
+
+    private static ClassMatcher getClassMatcher(Attributes attributes, String attribute, ClassMatcher defaultPatterns)
+    {
+        Object existing = attributes.getAttribute(attribute);
```

Comment:
Who is tracking what "Attributes" was used here?
So that the ee9 WebAppContext.setServer(Server) can use the appropriate `Attributes` objects to fetch the same ones?

It seems that callers of the methods on this new class will always use `Server`, right?

Reply:
The callers are primarily XML files and the PR has ones that call both with Server and with Environment.  None call this method directly, but via the types APIs